module-code/app/securesocial/core/providers/ChatWorkProvider.scala

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2017 Takeshi Kimura
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package securesocial.core.providers
+
+import play.api.{ Configuration, Environment }
+import play.api.libs.ws.WSAuthScheme
+import securesocial.core._
+import securesocial.core.services.{ CacheService, HttpService, RoutesService }
+
+import scala.concurrent.{ ExecutionContext, Future }
+import ChatWorkProvider._
+
+class ChatWorkOAuth2Client(
+    httpService: HttpService, settings: OAuth2Settings)(implicit executionContext: ExecutionContext) extends OAuth2Client.Default(httpService, settings)(executionContext) {
+  override def exchangeCodeForToken(code: String, callBackUrl: String, builder: OAuth2InfoBuilder): Future[OAuth2Info] = {
+    val params = Map(
+      OAuth2Constants.GrantType -> Seq(OAuth2Constants.AuthorizationCode),
+      OAuth2Constants.Code -> Seq(code),
+      OAuth2Constants.RedirectUri -> Seq(callBackUrl)
+    ) ++ settings.accessTokenUrlParams.mapValues(Seq(_))
+    httpService.url(settings.accessTokenUrl)
+      .withAuth(settings.clientId, settings.clientSecret, WSAuthScheme.BASIC)
+      .post(params).map(builder)
+  }
+
+}
+/**
+ * A ChatWork provider
+ *
+ */
+class ChatWorkProvider(routesService: RoutesService,
+  cacheService: CacheService,
+  client: OAuth2Client)(implicit val configuration: Configuration, val playEnv: Environment)
+    extends OAuth2Provider.Base(routesService, client, cacheService) {
+  override val id = ChatWorkProvider.ChatWork
+
+  override def fillProfile(info: OAuth2Info): Future[BasicProfile] = {
+    val accessToken = info.accessToken
+    client.httpService.url(ChatWorkProvider.Api).withHeaders("Authorization" -> s"Bearer $accessToken").get().map { response =>
+      response.status match {
+        case 200 =>
+          val data = response.json
+          val account_id = (data \ AccountId).asOpt[Int].getOrElse(UnknownUserId).toString()
```

Comment:
This line should be:
```
val accountId = (data \ AccountId).as[String]
```

* It seems like `account_id` in the response always exists, so we don't have to use `asOpt`, but `as`
* The type of `BasicProfile#userId` is `String`, so use `as[String]`, not `toString`
* In Scala, variable names should be camelCase

The reason why the type of `BasicProfile#userId`is `String` is that this library is used for many Providers such as Facebook and Slack, and some of them use String as their IDs. e.g. Slack

Reply:
let me clarify data \ AccountId. This instance is a part of the JSON and its value may have integer value itself (not string). In the JSON, account_id is coded in integer. so more relevant code is `val accountId = (data \ AccountId).as[Int].toString()`. toString is necessary.

Reply:
I tried your code `val accountId = (data \ AccountId).as[String]` and it failed with runtime error.

Reply:
see example JSON of /me.

http://developer.chatwork.com/ja/endpoint_me.html